File: kefiya/public/js/controllers/fints_interactive.js
// Copyright (c) 2019, jHetzer and contributors
// For license information, please see license.txt

frappe.provide("kefiya.interactive");

kefiya.interactive = {
	updateQueue: Promise.resolve(),
	progressState: {},

	enqueueUpdate: function(fn, delay) {
		this.updateQueue = this.updateQueue
			.then(() => {
				// call the function
				return Promise.resolve(fn());
			})
			.then(() => {
				if (!delay) {
					return;
				}

				// delay the next update only when explicitly requested
				return new Promise((resolve) => setTimeout(resolve, delay));
			});

		return this.updateQueue;
	},

	progressbar: function(frm) {
		this.progressState[frm.doc.name] = {
			progress: 0,
			completed: false,
		};

		frappe.realtime.off("fints_progressbar");
		frappe.realtime.on("fints_progressbar", function(data) {
			if(data.docname === frm.doc.name) {
				const state = kefiya.interactive.progressState[frm.doc.name] || {
					progress: 0,
					completed: false,
				};

				if (state.completed && data.progress < 100) {
					return;
				}

				if (data.progress < state.progress) {
					return;
				}

				state.progress = data.progress;
				state.completed = data.progress >= 100;
				kefiya.interactive.progressState[frm.doc.name] = state;

				kefiya.interactive.enqueueUpdate(() => {
					// One persistent window collecting every message, instead
					// of show_progress/hide_progress flashing a separate one
					// per step and per account.
					kefiya.progress.update(
						data.message, data.progress, data.docname
					);

					if(data.reload && data.reload === true) {
						// reload with short delay, to have progress update come active before
						kefiya.interactive.enqueueUpdate(() => {
							frm.reload_doc()
						}, 500);
					}
				}, 0);
			}
		});

		/**
		 * Handle TAN interaction requirements for several steps:
		 * 1. TAN Mode selection
		 * 1.2. TAN Medium selection (if available)
		 * 2. Later User Interaction after Push TAN fulfillment or TAN entered
		 */
		frappe.realtime.off("fints_tan_interaction_required");
		frappe.realtime.on("fints_tan_interaction_required", async function(data) {
			if(data.docname !== frm.doc.name) {
				return;
			}

			// "Freigabe erforderlich" named no bank and no account. During a
			// collective run one access after the other stops for a release
			// and the box looked identical every time, so the user had to
			// guess which banking app to open. The payload now carries the
			// access it belongs to.
			const title = data.account_label
				? __("Verification required") + " – " + data.account_label
				: __("Verification required");

			await kefiya.interactive.enqueueUpdate(() => {
				// Step out of the way for the TAN prompt, but keep the session
				// log: hide() retains the entries and the next update()
				// restores the window with its history intact.
				kefiya.progress.update(title, 0);
				kefiya.progress.hide();
			});

			let fields = [];

			if (data.possible_tan_modes) {
				fields.push(
					{
						fieldtype: "Select",
						fieldname: "tan_mode",
						label: __("TAN Mode"),
						options: data.possible_tan_modes,
						reqd: 1,
					},
				);

				if (data.possible_tan_mediums) {
					// if mediums are available for the selected mode, keep previously selected mode read-only
					fields[0].default = data.possible_tan_modes[0];
					fields[0].read_only = 1;

					fields.push(
						{
							fieldtype: "Select",
							fieldname: "tan_medium",
							label: __("TAN Medium"),
							options: data.possible_tan_mediums,
							reqd: 1,
						},
					);
				}
			}

			// TAN Mode selection
			if (data.tan_required || data.mfa_required) {
				if (data.possible_tan_modes && !fields[0].default) {
					fields[0].default = data.possible_tan_modes[0];
					fields[0].read_only = 1;
				}

				// if we're on confirmation already, keep previously selected mode read-only
				if (data.possible_tan_mediums) {
					const disableIdx = data.possible_tan_mediums ? 1 : 0;
					fields[disableIdx].default = data.possible_tan_mediums[disableIdx];
					fields[disableIdx].read_only = 1;
				}

				// User Interaction requirement for 2FA
				fields.push(
					{
						fieldtype: "Check",
						fieldname: "mfa_confirmation",
						label: __("Confirm MFA"),
						default: 1,
						hidden: 1,
					},
				);

				if (data.tan_required) {
					fields.push(
						{
							fieldtype: "Data",
							fieldname: "tan",
							label: __("TAN"),
							reqd: 1,
						},
					);
				} else if (data.mfa_required) {
					fields.push(
						{
							fieldtype: "HTML",
							fieldname: "waiting",
							label: __("Waiting for Interaction"),
							options: __("Follow the instructions on your banking app or device."),
						},
					);
				}
			}

			// Prepended only now: everything above addresses the fields by
			// index (fields[0] is the mode, fields[1] the medium), so an entry
			// inserted ahead of them earlier would silently make the wrong
			// field read-only.
			if (data.account_detail) {
				fields.unshift({
					fieldtype: "HTML",
					fieldname: "kefiya_tan_context",
					options: '<div class="text-muted small" style="margin-bottom:8px">'
						+ frappe.utils.escape_html(data.account_detail)
						+ "</div>",
				});
			}

			const dialog = frappe.prompt(
				fields,
				(values) => {
					frappe.call({
						method: "kefiya.utils.client.resolve_tan_interaction",
						args: {
							// The login the run belongs to, which is this form only
							// on the Kefiya Login screen. A transfer's box used to
							// answer against its own KEF-TRF-... name.
							fints_login: data.fints_login || frm.doc.name,
							values: { ...data, ...values },
						},
					});
				},
				title
			);

			// Straight into the TAN field: the mode and the medium above it
			// are read-only, so the only thing to do here is type. The rule
			// itself lives in bank_refresh.js, which owns the other TAN box.
			if (kefiya.focus_tan_field) kefiya.focus_tan_field(dialog);
		});
	},
}

File: kefiya/utils/fints_interactive.py
# -*- coding: utf-8 -*-
# Copyright (c) 2026, Phamos GmbH and contributors
# For license information, please see license.txt

"""Everything the bank dialog says to the person waiting for it.

Progress and the release prompt, and nothing else. It sat inside
fints_controller.py, which is two thousand lines of FinTS protocol -- a file
nobody opens to find out what a dialog box says. The class never touched the
protocol; it only ever published realtime events, so it moves out whole.

fints_controller.py imports it back under its old name, so every existing
caller is unaffected.
"""

import frappe
from frappe import _

from kefiya.utils.fints_masking import mask_iban


class FinTSInteractive:
    def __init__(self, configuration):
        if not configuration:
            self.docname = None
            self.enabled = False
        else:
            self.docname = configuration["docname"]
            self.enabled = configuration["enabled"]
        # Which Kefiya Login this is about, as opposed to which form the user
        # is looking at. They are the same thing on the Kefiya Login form and
        # nowhere else: a transfer passes its own docname as the UI scope, so
        # everything that read the login out of `docname` was reading a
        # KEF-TRF-... name. The account context came back empty (the box named
        # no bank and no account, on the one screen where money moves), and a
        # box that tried to answer resolved against a document that is not a
        # login at all.
        self.fints_login = None
        self.progress = 0
        # Resolved on the first prompt, not here: a controller that never asks
        # for a TAN -- the scheduled run, most of the time -- should not pay a
        # lookup for a dialog it will not open.
        self._account_context_cache = None

    def login_name(self):
        """The Kefiya Login this run belongs to.

        Falls back to the UI scope, which is the same name on the Kefiya Login
        form -- so a caller that never sets the login behaves exactly as
        before.
        """
        return self.fints_login or self.docname

    def set_interactive_mode(self, enable):
        """Turn on/off interactive mode.

        :param enable: Turn on/off interactive mode
        :type enable: bool
        :return: None
        """
        self.enabled = enable

    def get_interactive_mode(self):
        """Get interactive mode.

        :return: bool
        """
        return self.enabled

    def show_progress_realtime(self, message, progress, reload=False):
        """Show a progressbar on client side.

        :param message: Message to display under the bar
        :param progress: 0 - 100
        :param reload: Reload the doc form, , defaults to False
        :type message: str
        :type progress: int
        :type reload: bool, optional
        :return: None
        """
        if self.enabled:
            frappe.publish_realtime(
                "fints_progressbar", {
                    "progress": progress,
                    "docname": self.docname,
                    "message": message,
                    "reload": reload
                }, user=frappe.session.user)

    def request_tan_mechanism(self, possible_tan_modes=None, possible_tan_mediums=None):
        """Request tan mechanism from user.

        :param possible_tan_modes: List of tan mechanisms
        :type mechanisms: list
        :return: None
        """
        self.request_tan_prompt(possible_tan_modes, possible_tan_mediums)

    def request_tan(self, possible_tan_modes=None, possible_tan_mediums=None, challenge=None):
        """Request a TAN from user

        :return: None
        """
        self.request_tan_prompt(possible_tan_modes, possible_tan_mediums,
                                request_tan=True, challenge=challenge)

    def request_mfa_confirmation(self, possible_tan_modes=None, possible_tan_mediums=None, challenge=None):
        """Request a solved MFA challenge from the user

        :return: None
        """
        self.request_tan_prompt(possible_tan_modes, possible_tan_mediums,
                                request_mfa_confirmation=True,
                                challenge=challenge)

    def account_context(self):
        """Which bank and which account this prompt is about.

        "Freigabe erforderlich -- TAN" said nothing else. During a collective
        fetch that is a dialog with no sender: a dozen accesses run one after
        the other, each may stop for a release, and the box on screen names
        none of them. The user then has to guess which banking app to open,
        and a release given in the wrong app is a release the waiting dialog
        never gets.

        So the prompt carries the access it belongs to. The IBAN is masked to
        its last four digits -- enough to tell two accounts of the same company
        apart, and no account number on a screen that may be shared.

        :return: dict with bank, account_name, iban (masked), label, detail
        """
        if self._account_context_cache is not None:
            return self._account_context_cache

        context = {}
        try:
            bank_account, iban = frappe.db.get_value(
                "Kefiya Login", self.login_name(),
                ["bank_account", "account_iban"]) or (None, None)

            bank = account_name = None
            if bank_account:
                bank, account_name, account_iban = frappe.db.get_value(
                    "Bank Account", bank_account,
                    ["bank", "account_name", "iban"]) or (None, None, None)
                # The login's own IBAN is the more specific one -- a login may
                # point at an account record that carries none.
                iban = iban or account_iban

            parts = [p for p in (bank, account_name or self.login_name()) if p]
            context = {
                "bank": bank,
                "account_name": account_name,
                "iban": mask_iban(iban) if iban else None,
                # Short enough for a dialog title, which truncates.
                "account_label": " · ".join(parts),
                # The full line, for a field inside the dialog.
                "account_detail": " · ".join(
                    parts + ([mask_iban(iban)] if iban else [])),
            }
        except Exception:
            # A prompt without its heading is worse than the old one; a prompt
            # that never appears because the heading could not be looked up is
            # worse still. The release matters, the label does not -- which is
            # why even the logging of the failure is guarded: an Error Log that
            # cannot be written must not be the reason a bank dialog is left
            # standing without anyone being asked to release it.
            try:
                frappe.log_error(
                    title="Kefiya: TAN prompt context could not be resolved",
                    message=frappe.get_traceback(),
                    reference_doctype="Kefiya Login",
                    reference_name=self.docname,
                )
            except Exception:
                pass

        self._account_context_cache = context
        return context

    def request_tan_prompt(self, possible_tan_modes, possible_tan_mediums=None, *, request_tan=False, request_mfa_confirmation=False, challenge=None):
        """Request tan mechanism from user.

        :param possible_tan_modes: List of tan mechanisms
        :type mechanisms: list
        :return: None
        """
        if self.enabled:
            params = {
                        "docname": self.docname,
                        # The login to answer against, which is NOT the form
                        # the user is looking at whenever money is moving.
                        # Every box resolves with this; without it a transfer's
                        # box answered against a KEF-TRF-... name.
                        "fints_login": self.login_name(),
                        "possible_tan_modes": possible_tan_modes,
                        "possible_tan_mediums": possible_tan_mediums,
                    }

            # Say which access is asking. Three prompts read this event -- the
            # form, the cockpit refresh and the outgoing-payments block -- and
            # each of them showed a bare "Verification required" before.
            params.update(self.account_context())

            # The challenge itself, where the bank sent one. A photoTAN image
            # travels as base64 in this payload: a few kilobytes, and the only
            # way the picture reaches the screen that has to show it.
            if challenge:
                params["challenge"] = challenge

            if request_tan:
                params["tan_required"] = True

            elif request_mfa_confirmation:
                params["mfa_required"] = True

            frappe.publish_realtime("fints_tan_interaction_required", params, user=frappe.session.user)
